pkg/compiler/binary_expr_test.go

```diff
@@ -266,83 +270,134 @@ func TestBinaryExprs(t *testing.T) {
-			src := fmt.Sprintf(srcTmpl, "true", cond, "", "")
-			evalLayout(t, src, res, e)
-		})
-		t.Run("InConditionWithElse", func(t *testing.T) {
-			src := fmt.Sprintf(srcTmpl, "true", cond, " else {", "}")
-			evalLayout(t, src, res, e)
+
+	return append(testCases, funcTestCase{
+		Name:   cond,
+		Result: res,
+	})
+}
+
+func runBooleanCases(t *testing.T, testCases []funcTestCase, src string) {
+	ne, di, err := compiler.CompileWithOptions("file.go", strings.NewReader(src), nil)
+	require.NoError(t, err)
+
+	for i, tc := range testCases {
+		v := vm.New()
+		t.Run(tc.Name, func(t *testing.T) {
+			t.Run("AsExpression", func(t *testing.T) {
+				v.Istack().Clear()
+				v.Estack().Clear()
+				invokeMethod(t, fmt.Sprintf("F%d_expr", i), ne.Script, v, di)
+				runAndCheck(t, v, tc.Result)
+			})
+			t.Run("InCondition", func(t *testing.T) {
+				v.Istack().Clear()
+				v.Estack().Clear()
+				invokeMethod(t, fmt.Sprintf("F%d_cond", i), ne.Script, v, di)
+				runAndCheck(t, v, tc.Result)
+			})
+			t.Run("InConditionWithElse", func(t *testing.T) {
+				v.Istack().Clear()
+				v.Estack().Clear()
+				invokeMethod(t, fmt.Sprintf("F%d_else", i), ne.Script, v, di)
+				runAndCheck(t, v, tc.Result)
+			})
 		})
 	}
 }
 
 // TestBooleanExprs enumerates a lot of possible combinations of boolean expressions
 // and tests if the result matches to that of Go.
 func TestBooleanExprs(t *testing.T) {
-	t.Skip() // FIXME this test takes more than 2 minutes to complete
+	header := `package foo
+	var s = "str"
+	var v = 9
+	`
+
+	srcBuilder := bytes.NewBuffer([]byte(header))
 
-	e := preparePackageLayout(t)
+	var testCases []funcTestCase
 
 	trueExpr := []string{"true", "v < 10", "v <= 9", "v > 8", "v >= 9", "v == 9", "v != 8", `s == "str"`}
 	falseExpr := []string{"false", "v > 9", "v >= 10", "v < 9", "v <= 8", "v == 8", "v != 9", `s == "a"`}
 	t.Run("Single", func(t *testing.T) {
 		for _, s := range trueExpr {
-			t.Run(s, getBoolExprTestFunc(true, s, e))
+			testCases = addBoolExprTestFunc(testCases, srcBuilder, true, s)
 		}
 		for _, s := range falseExpr {
-			t.Run(s, getBoolExprTestFunc(false, s, e))
+			testCases = addBoolExprTestFunc(testCases, srcBuilder, false, s)
 		}
+		runBooleanCases(t, testCases, srcBuilder.String())
 	})
 
 	type arg struct {
 		val bool
 		s   string
 	}
-	t.Run("Combine", func(t *testing.T) {
-		var double []arg
-		for _, e := range trueExpr {
-			double = append(double, arg{true, e + " || false"})
-			double = append(double, arg{true, e + " && true"})
-		}
-		for _, e := range falseExpr {
-			double = append(double, arg{false, e + " && true"})
-			double = append(double, arg{false, e + " || false"})
-		}
+	var double []arg
+	for _, e := range trueExpr {
+		double = append(double, arg{true, e + " || false"})
+		double = append(double, arg{true, e + " && true"})
+	}
+	for _, e := range falseExpr {
+		double = append(double, arg{false, e + " && true"})
+		double = append(double, arg{false, e + " || false"})
```

Comment:
`e` always on the left?

Reply:
It was this way before, wiil add new tests.

Reply:
Actually, I remember why it was this way: we already have `false` and `true` in `falseExpr` and `trueExpr` respectively, so this was enough for testing short-circuit logic.